core-rust/core-api-server/src/core_api/handlers/stream_proofs.rs

```diff
@@ -0,0 +1,211 @@
+            min: 1,
+            default: 20,
+            max: 100,
+        },
+    )
+    .map_err(|err| err.into_response_error("max_page_size"))?;
+
+    let continue_from_state_version =
+        extract_continuation_token::<StateVersion>(request.continuation_token)
+            .map_err(|err| err.into_response_error("continuation_token"))?;
+
+    let database = state.state_manager.database.read_current();
+
+    use models::StreamProofsFilter::*;
+    let mut proofs_iter = match filter {
+        StreamProofsFilterAll { from_state_version } => iterate_all_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_state_version(&database, from_state_version)?,
+        ),
+        StreamProofsFilterNewEpochs { from_epoch } => iterate_end_of_epoch_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_epoch(&mapping_context, &database, from_epoch)?,
+        ),
+        StreamProofsFilterProtocolUpdateInitializations { from_state_version } => {
+            iterate_protocol_update_initialization_proofs(
+                &database,
+                continue_from_state_version,
+                extract_from_state_version(&database, from_state_version)?,
+            )
+        }
+        StreamProofsFilterProtocolUpdateExecution {
+            from_state_version,
+            protocol_version,
+        } => iterate_protocol_update_execution_proofs(
+            &database,
+            continue_from_state_version,
+            extract_from_state_version(&database, from_state_version)?,
+            protocol_version,
+        ),
+    }?;
+
+    let (page, continuation_token) = to_api_page(
+        &mut proofs_iter,
+        page_size,
+        |proof| handlers::to_api_ledger_proof(&mapping_context, proof),
+        |proof| proof.ledger_header.state_version,
+    )?;
+
+    Ok(Json(models::StreamProofsResponse {
+        page,
+        continuation_token,
+    }))
+}
+
+fn iterate_all_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_state_version: StateVersion,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let start = optional_max(from_state_version, continue_from_state_version);
+
+    Ok(database.get_proof_iter(start))
+}
+
+fn iterate_end_of_epoch_proofs<'a>(
+    database: &'a StateManagerDatabase,
+    continue_from_state_version: Option<StateVersion>,
+    from_epoch: Epoch,
+) -> Result<
+    Box<dyn Iterator<Item = LedgerProof> + 'a>,
+    ResponseError<models::StreamProofsErrorDetails>,
+> {
+    let continuation_next_epoch = match continue_from_state_version {
+        Some(state_version) => match database.get_proof_iter(state_version).next() {
+            Some(proof) => Some(proof.ledger_header.epoch.next().unwrap()),
+            None => Err(client_error("continuation_token is not valid"))?,
```

Comment:
here - yes, this means malformed continuation token

but in general, it can happen to a valid token, when elements are deleted from the end of the collection. So maybe an empty return is more appropriate than error?

On a related note: I can imagine an impl in which the continuation token for epoch-proof-listing is `continue_from_epoch: Epoch` (instead of state version). But we don't have to pursue that if too cumbersome for "Rust typing" reasons.

Reply:
> but in general, it can happen to a valid token, when elements are deleted from the end of the collection. So maybe an empty return is more appropriate than error?

Agreed in general, but I think either's fine here. It should never happen in practice, and is a user error, so I think it's fine to 400. This code is all just for converting state version to next epoch anyway.

> On a related note: I can imagine an impl in which the continuation token for epoch-proof-listing is continue_from_epoch: Epoch (instead of state version). But we don't have to pursue that if too cumbersome for "Rust typing" reasons.

Yeah I considered making the continuation token the epoch in this case....

I was torn, but in the end, decided that viewing proofs as having a primary key of state version, regardless of which index we read from was conceptually nicer. I'm honestly not totally sure if that was the right choice, but it was quite arbitrary and honestly I don't think it matters _too_ much. So would be tempted to leave unless you feel strongly.

Reply:
> viewing proofs as having a primary key of state version, regardless of which index we read from was conceptually nicer.

From one point of view ("the exact index used is only an impl detail"), I agree.
But then again, the contents of ContinuationToken is (very explicitly) an opaque impl detail too... It would clearly be more convenient for the DB layer if ContinuationToken contained the epoch directly (i.e. it would save an extra DB roundtrip [which I point out for conceptual reasons, not performance]). But I understand it is less convenient to the controller (for Rust typing reasons, or code reusability reasons).

So yeah I don't feel strongly in this case, we can keep "whatever is easier" here, and just allow "DB-driven ContinuationToken contents" in general. After all, this _is_ an impl detail, and our "API conventions" don't even need to force anything here.